apps/els_lsp/src/els_parser.erl

```diff
@@ -568,14 +569,84 @@ subtrees(Tree, record_field) ->
       V ->
        [V]
     end];
+subtrees(Tree, record_type) ->
+  NameNode = erl_syntax:record_type_name(Tree),
+  [ skip_record_field_atom(NameNode)
+  , erl_syntax:record_type_fields(Tree)
+  ];
+subtrees(Tree, record_type_field) ->
+  NameNode = erl_syntax:record_type_field_name(Tree),
+  [ skip_record_field_atom(NameNode)
+  , [erl_syntax:record_type_field_type(Tree)]
+  ];
+subtrees(Tree, user_type_application) ->
+  NameNode = erl_syntax:user_type_application_name(Tree),
+  [ skip_record_field_atom(NameNode)
+  , erl_syntax:user_type_application_arguments(Tree)
+  ];
+subtrees(Tree, type_application) ->
+  NameNode = erl_syntax:type_application_name(Tree),
+  [ skip_type_name_atom(NameNode)
+  , erl_syntax:type_application_arguments(Tree)
+  ];
 subtrees(Tree, attribute) ->
-  case erl_syntax:attribute_arguments(Tree) of
-    none -> [];
-    [_ | RestArgs] -> [RestArgs]
-  end;
+  NameNode = erl_syntax:attribute_name(Tree),
+  AttrName = case erl_syntax:type(NameNode) of
+               atom -> erl_syntax:atom_value(NameNode);
+               _ -> NameNode
+             end,
+  Args = case erl_syntax:attribute_arguments(Tree) of
+           none -> [];
+           Args0 -> Args0
+         end,
+  attribute_subtrees(AttrName, Args);
 subtrees(Tree, _) ->
   erl_syntax:subtrees(Tree).
 
+%% Note: In erl_parse AST the arguments of a wild attribute are represented as
+%% plain terms. To make response consistent `attribute_arguments/1' returns the
+%% abstract format of those terms. However `erl_syntax' doesn't know that
+%% `callback', `spec', `type' and `opaque' are not wild attributes and have
+%% their arguments in abstract format already. So `erl_syntax' returns the AST
+%% of the AST for these attributes. To fix this we need to convert them back
+%% with `concrete/1' to be able to properly traverse them.
+-spec attribute_subtrees(atom() | tree(), [tree()]) -> [[tree()]].
+attribute_subtrees(AttrName, [Mod])
+  when AttrName =:= module;
+       AttrName =:= behavior;
+       AttrName =:= behaviour ->
+  [skip_record_field_atom(Mod)];
+attribute_subtrees(record, [_RecordName, FieldsTuple]) ->
+  [[FieldsTuple]];
+attribute_subtrees(import, [Mod, Imports]) ->
+  [ skip_record_field_atom(Mod)
+  , [Imports]];
+attribute_subtrees(define, [_Name | Definition]) ->
+  %% The definition can contain commas, in which case it will look like as if
+  %% the attribute would have more than two arguments. Eg.: `-define(M, a, b).'
+  [Definition];
+attribute_subtrees(AttrName, _)
+  when AttrName =:= include;
+       AttrName =:= include_lib ->
+  [];
+attribute_subtrees(AttrName, [Arg])
+  when AttrName =:= callback;
+       AttrName =:= spec ->
+  {_FA, DefinitionClauses} = erl_syntax:concrete(Arg),
+  [DefinitionClauses];
+attribute_subtrees(AttrName, [Arg])
+  when AttrName =:= type;
+       AttrName =:= opque ->
```

Comment:
Typo: opque

Reply:
thanks, I realize I havent added tests for opaque and callback, will do